### PR TITLE
HRTWorkQueue: avoid scheduling drift

### LINKED ISSUE #68: The DF timer includes drift
@mcharleb @jywilson If you ask the DF for a callback at a 4000 us interval you're getting 4000 us + scheduling overhead, not 4000 us on average, which is a problem for us. I will look with @bkueng into this, but if you have availability it would be great if you could propose a fix.

### PR DESCRIPTION
**Pull request by bkueng**
This should fix #68.

It would also be a good idea to check if there's too much work & scheduling falls behind.

@LorenzMeier can you test if this improves things?

btw `make run` shows some failed tests, but the same happens on master (not reported as error in travis).

### CODE REVIEW

File: framework/src/DriverFramework.cpp [hunk 1]
```diff
@@ -602,36 +602,34 @@ void HRTWorkQueue::process(void)
...
-		DF_LOG_DEBUG("HRTWorkQueue::process waiting for work (%" PRIi64 ")", diff);
-		// Wait until next expiry or until a new item is rescheduled
-		pthread_cond_timedwait(&g_reschedule_cond, &g_hrt_lock, &ts);
+		if (next > offsetTime()) {
+			// pthread_cond_timedwait uses absolute time
+			ts = offsetTimeToAbsoluteTime(next);
+			DF_LOG_DEBUG("HRTWorkQueue::process waiting for work (%" PRIi64 ")", next - offsetTime());
+			// Wait until next expiry or until a new item is rescheduled
+			pthread_cond_timedwait(&g_reschedule_cond, &g_hrt_lock, &ts);
+		}
```
Comment (julianoes) [inline framework/src/DriverFramework.cpp]: And `else`? Shouldn't it get scheduled straightaway?

File: framework/src/DriverFramework.cpp [hunk 2]
```diff
@@ -577,22 +577,22 @@ void HRTWorkQueue::process(void)
...
+
 		while ((!m_exit_requested) && (idx != nullptr)) {
 			DF_LOG_DEBUG("HRTWorkQueue::process work exists");
-			now = offsetTime();
 			unsigned int index;
 			m_work_list.get(idx, index);
 
 			if (index < g_work_items->size()) {
-				WorkItem *item = nullptr;
+				WorkItem *item;
```
Comment (julianoes) [inline framework/src/DriverFramework.cpp]: Supposedly the RPi compiler wants this initialized to `nullptr`.

File: framework/src/DriverFramework.cpp (hunk 1 above)
Comment (bkueng) [inline framework/src/DriverFramework.cpp]: That's exactly what it does, because it stays in the loop

Comment (julianoes) [inline framework/src/DriverFramework.cpp]: Ah yes, makes sense.

**Comment by LorenzMeier** [comment]
@mcharleb @jywilson Could you have a look?

**Comment by julianoes** [comment]
Thanks @bkueng!

I gave this a test on the Snapdragon:

**Without fix:**
```
[08500/00]  15:24.647  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:25.765  HAP:60:average interval: 1118  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:26.885  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:28.005  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:29.123  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:30.243  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:31.363  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:32.482  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:33.600  HAP:60:average interval: 1117  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:34.720  HAP:60:average interval: 1120  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:35.840  HAP:60:average interval: 1120  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:36.960  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:38.080  HAP:60:average interval: 1119  0496  df_mpu9250_wrapper.cpp
[08500/00]  15:39.200  HAP:60:average interval: 1120  0496  df_mpu9250_wrapper.cpp
```

**With this PR:**
```
[08500/00]  23:06.036  HAP:60:average interval: 998  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:07.035  HAP:60:average interval: 999  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:08.035  HAP:60:average interval: 999  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:09.033  HAP:60:average interval: 998  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:10.032  HAP:60:average interval: 998  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:11.032  HAP:60:average interval: 999  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:12.031  HAP:60:average interval: 999  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:13.030  HAP:60:average interval: 998  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:14.028  HAP:60:average interval: 998  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:15.027  HAP:60:average interval: 999  0496  df_mpu9250_wrapper.cpp
[08500/00]  23:16.027  HAP:60:average interval: 999  0496  df_mpu9250_wrapper.cpp
```

Seems like it's working as it should.

File: framework/src/DriverFramework.cpp [hunk 3]
```diff
@@ -602,36 +601,34 @@ void HRTWorkQueue::process(void)
 					item->updateStats(now);
 
 					// reschedule work
-					item->m_queue_time = offsetTime();
+					item->m_queue_time += item->m_delay_usec;
```
Comment (mcharleb) [inline framework/src/DriverFramework.cpp]: m_queue_time used to be the offset time that the work item was (re)queued. Now it is being set to the delay interval.

Comment (mcharleb) [inline framework/src/DriverFramework.cpp]: Much better! Good work!

**Pull request merged**